compiler/rustc_ast/src/tokenstream.rs

```diff
@@ -410,8 +410,19 @@ impl TokenStream {
         t1.next().is_none() && t2.next().is_none()
     }
 
-    pub fn map_enumerated<F: FnMut(usize, &TokenTree) -> TokenTree>(self, mut f: F) -> TokenStream {
-        TokenStream(Lrc::new(self.0.iter().enumerate().map(|(i, tree)| f(i, tree)).collect()))
+    pub fn map_enumerated_owned<F: FnMut(usize, TokenTree) -> TokenTree>(
```

Comment:
Can you add a brief doc comment explaining what this function does? It's sufficiently complicated that it's not self-explanatory.